src/lib/logs/log_forwarder.py

```diff
@@ -12,134 +12,75 @@
+):
     if not LOGS_SUBSCRIPTION_PROJECT or not LOGS_SUBSCRIPTION_ID:
         raise Exception(
-            "Cannot start pubsub streaming pull - GCP_PROJECT or LOGS_SUBSCRIPTION_ID are not defined")
+            "Cannot start pubsub pulling - GCP_PROJECT or LOGS_SUBSCRIPTION_ID are not defined"
+        )
+
+    # Each process starts later than the previous one
+    await asyncio.sleep(process_number * LOG_PROCESS_STARTUP_DELAY_SECONDS)
 
     sfm_queue = Queue(MAX_SFM_MESSAGES_PROCESSED)
-    subscriber_client = pubsub.SubscriberClient()
-    subscription_path = subscriber_client.subscription_path(LOGS_SUBSCRIPTION_PROJECT, LOGS_SUBSCRIPTION_ID)
-
-    # Open worker threads to process logs from PubSub queue and ingest them into DT
-    for i in range(0, PROCESSING_WORKERS):
-        threading.Thread(target=pull_and_flush_logs_forever,
-                         args=(f"Worker-{i}", sfm_queue, subscriber_client, subscription_path,),
-                         name=f"worker-{i}").start()
-
-    # Create loop with a timer to gather self monitoring metrics and send them to GCP (if enabled)
-    create_sfm_loop(sfm_queue, logging_context, instance_metadata)
-
-
-def pull_and_flush_logs_forever(worker_name: str,
-                                sfm_queue: Queue,
-                                subscriber_client: SubscriberClient,
-                                subscription_path: str):
-    logging_context = LoggingContext(worker_name)
-    worker_state = WorkerState(worker_name)
-    pull_request = PullRequest()
-    pull_request.max_messages = PROCESSING_WORKER_PULL_REQUEST_MAX_MESSAGES
-    pull_request.subscription = subscription_path
+    log_integration_service = LogIntegrationService(sfm_queue)
+    await log_integration_service.update_gcp_client(logging_context)
+
+    tasks = []
+
+    update_token = asyncio.create_task(log_integration_service.keep_gcp_token_updated(logging_context))
+
+    tasks.append(update_token)
+    for worker_number in range(0, PROCESSING_WORKERS):
+        worker_task = asyncio.create_task(
+            pull_and_push_logs_forever(process_number, worker_number, log_integration_service)
+        )
+        tasks.append(worker_task)
+
+    sfm_task = asyncio.create_task(create_sfm_loop(sfm_queue, logging_context, instance_metadata))
+    tasks.append(sfm_task)
+
+    await asyncio.gather(*tasks)
+
+
+async def pull_and_push_logs_forever(
+    process_number: int, worker_number: int, log_integration_service: LogIntegrationService
+):
+    logging_context = LoggingContext(f"Process-{process_number}", f"Worker-{worker_number}")
     logging_context.log(f"Starting processing")
     while True:
         try:
-            perform_pull(worker_state, sfm_queue, subscriber_client, subscription_path, pull_request)
+            # Pull logs from pub/sub, process them and create batches out of them to send to Dynatrace
+            log_batches, ack_ids = await log_integration_service.perform_pull(logging_context)
```

Comment:
These ack_ids are only from the erroneous messages, as I understand (and we are adding later the ids from the sent messages). So we can also be more descriptive with the naming.